src/devices/sound/nds_sound.cpp

```diff
@@ -0,0 +1,829 @@
+{
+	return m_channel[(offset >> 2) & 0xf].control;
+}
+
+
+//-------------------------------------------------
+//  channel_w - write sound channel register
+//-------------------------------------------------
+
+void nds_sound_device::channel_w(offs_t offset, u32 data, u32 mem_mask)
+{
+	m_channel[(offset >> 2) & 0xf].write(offset & 3, data, mem_mask);
+}
+
+
+//-------------------------------------------------
+//  channel_t::write - write sound channel register
+//-------------------------------------------------
+
+void nds_sound_device::channel_t::write(offs_t offset, u32 data, u32 mem_mask)
+{
+	const u32 old = control;
+	switch (offset & 3)
+	{
+	case 0: // Control/Status
+		COMBINE_DATA(&control);
+		if (BIT(old ^ control, 31))
+			keyon();
+		else
+			keyoff();
+
+		// reset hold flag
+		if (!playing && !hold())
+		{
+			sample = lfsr_out = 0;
+			output = loutput = routput = 0;
+		}
+		break;
+	case 1: // Source address
+		mem_mask &= 0x7ffffff;
+		COMBINE_DATA(&sourceaddr);
+		break;
+	case 2: // Frequency, Loopstart
+		if (ACCESSING_BITS_0_15)
+			freq = (freq & BIT(~mem_mask, 0, 16)) | (BIT(data & mem_mask, 0, 16));
+		if (ACCESSING_BITS_16_31)
+			loopstart = (loopstart & BIT(~mem_mask, 16, 16)) | (BIT(data & mem_mask, 16, 16));
+		break;
+	case 3: // Length
+		mem_mask &= 0x3fffff;
+		COMBINE_DATA(&length);
+		break;
+	}
+}
+
+
+//-------------------------------------------------
+//  channel_t::keyon - execute sound channel keyon
+//-------------------------------------------------
+
+void nds_sound_device::channel_t::keyon()
+{
+	if (!playing)
+	{
+		playing = true;
+		delay = format() == 2 ? 11 : 3; // 3 (11 for ADPCM) delay for playing sample
+		cur_bitaddr = cur_addr = 0;
+		cur_state = format() == 2 ? STATE_ADPCM_LOAD : (loopstart == 0 ? STATE_POST_LOOP : STATE_PRE_LOOP);
+		counter = 0x10000;
+		sample = 0;
+		lfsr_out = 0x7fff;
+		lfsr = 0x7fff;
+	}
+}
+
+
+//-------------------------------------------------
+//  channel_t::keyoff - execute sound channel keyoff
+//-------------------------------------------------
+
+void nds_sound_device::channel_t::keyoff()
+{
+	if (playing)
+	{
+		if (busy())
+			control &= ~(1 << 31);
+		if (!hold())
+		{
+			sample = lfsr_out = 0;
+			output = loutput = routput = 0;
+		}
+
+		playing = false;
+	}
+}
+
+
+//-------------------------------------------------
+//  channel_t::fetch - fetch sound sample
+//-------------------------------------------------
+
+void nds_sound_device::channel_t::fetch()
+{
+	if (playing)
+	{
+		// fetch samples
+		switch (format())
+		{
+		case 0: // PCM8
+			sample = s16(host->m_cache.read_byte(addr()) << 8);
+			break;
+		case 1: // PCM16
+			sample = host->m_cache.read_word(addr());
+			break;
+		case 2: // ADPCM
+			sample = cur_state == STATE_ADPCM_LOAD ? 0 : adpcm_out;
+			break;
+		case 3: // PSG or Noise
+			sample = 0;
+			if (psg) // psg
+				sample = duty() == 7 ? -0x8000 : cur_bitaddr < (7 - duty()) ? -0x8000 : 0x7fff;
```

Comment:
These conditional statements would also be more readable with the conditions in parentheses.